File: front/src/types/inbox.ts
export type InboxList = {
    count: number;
    results: Inbox[];
}

export type Inbox = {
    ID: string;
    Name: string;
    Timestamp: number;
    Response: InboxResponse;
    Requests: InboxRequest[];
    ObfuscateHeaderFields: string[];
    IsPrivate: boolean;
    OwnerID: string;
    Callbacks: InboxCallback[];
}

export type InboxCallback = {
    IsEnabled: boolean;
    IsDynamic: boolean;
    ToURL: string;
    Method: string;
    Headers: Record<string, string>;
    Body: string;
}

export type CallbackResponse = {
    URL: string;
	Method: string;
	Error: string;
    Code: number;
    CodeTemplate: string;
    Body: string;
    Headers: Record<string, string>;
}

export type InboxRequest = {
    ID: number;
    Timestamp: number;
    URI: string;
    Headers: Record<string, string>;
    Body: string;
    Host: string;
    RemoteAddr: string;
    Protocol: string;
    Method: string;
    ContentLength: number;
    CallbackResponses?: CallbackResponse[];
}

export type InboxResponse = {
    Code: number;
    CodeTemplate: string;
    Body: string;
    Headers: Record<string, string>;
    IsDynamic: boolean;
}

export type User = {
    ID: string;
    Name: string;
    Email: string;
    Organization: string;
    AvatarURL: string;
}

export type APIKeyList = {
    count: number;
    results: APIKey[];
}

export type APIKey = {
    ID: string;
    Name: string;
    OwnerID: string;
    APIKey: string;
    CreationDate: Date;
    ExpiryDate: Date | null;
    IsActive: boolean;
}

File: api/pkg/model/inbox.go
package model

import (
	"net/http"
	"time"

	"github.com/google/uuid"
)

const (
	ContentTypeHeader        = "Content-Type"
	DefaultContentTypeHeader = "application/json; charset=utf-8"
	DefaultBody              = "{}"
	InboxEntityName          = "Inbox"
)

type Inbox struct {
	ID                    uuid.UUID
	Name                  string     `dynamodbav:"alias"`
	Timestamp             int64      `dynamodbav:"unixTimestamp"`
	Response              Response   `dynamodbav:"resp"`
	Requests              []Request  `dynamodbav:"req"`
	ObfuscateHeaderFields []string   `dynamodbav:"ofuscate"`
	Callback              []Callback `dynamodbav:"Callback"`
	OwnerID               uuid.UUID  `dynamodbav:"OwnerID"`
	IsPrivate             bool       `dynamodbav:"IsPrivate"`
}

type Response struct {
	Code         int
	CodeTemplate string
	Body         string
	Headers      map[string]string
	IsDynamic    bool
}

type CallbackResponse Response

type Request struct {
	ID                int
	Timestamp         int64 `dynamodbav:"unixTimestamp"`
	URI               string
	Host              string
	RemoteAddr        string
	Protocol          string
	Headers           map[string][]string
	Method            string `dynamodbav:"httpMethod"`
	ContentLength     int64
	Body              string
	CallbackResponses []CallbackResponse
}

type Callback struct {
	IsEnabled bool `dynamodbav:"enabled"`
	IsDynamic bool `dynamodbav:"isDynamic"`
	// PrefixURL string            `dynamodbav:"prefixURL"`
	ToURL   string            `dynamodbav:"toURL"`
	Method  string            `dynamodbav:"method"`
	Headers map[string]string `dynamodbav:"headers"`
	Body    string            `dynamodbav:"body"`
}

func NewInbox() Inbox {
	id := uuid.New()
	return Inbox{
		ID:        id,
		Name:      id.String(),
		Timestamp: time.Now().UnixMilli(),
		Response: Response{
			Code:    http.StatusOK,
			Headers: map[string]string{ContentTypeHeader: DefaultContentTypeHeader},
			Body:    DefaultBody,
		},
		Requests:              []Request{},
		ObfuscateHeaderFields: []string{},
		Callback:              []Callback{},
		IsPrivate:             false,
		OwnerID:               uuid.UUID{},
	}
}

func NewCallback() Callback {
	return Callback{
		IsEnabled: false,
		IsDynamic: false,
		Method:    "",
		ToURL:     "",
		// PrefixURL: "",
		Headers: map[string]string{},
		Body:    "",
	}
}
